### ISSUE TITLE
How to properly manage scalar bars for multiple meshes?

### ISSUE THREAD

**Issue by math-artist** [pyvista/pyvista-support#99]
Hi,

I currently have this problem: I have many meshes that can be toggled visible or hidden, and toggled "colormap visible or hidden".

They all have their own scalar bar associated with them, most of the time it's the same, but sometimes it's not.

I cannot find a proper way to delete the scalar_bar actor if I need to delete the mesh actor. I normally use the remove_actor("mesh_name") method to hide my meshes, and I can't figure out a clean way to do it on their associated color_bars.

Is there a way to name or refer to the scalar_bars by id, or a way to simply delete all scalar_bars without deleting the other actors? It seems that I can only delete the latest one.

Or, is there a way to always overwrite the same scalar bar, so that it get rids of this problem?

Right now, I can remove the scalar bar if I have only one displayed, and I didn't find how to detect that there is already a scale bar displayed.

**Comment by banesullivan** [comment]
Hey, @math-artist the scalar bars are managed in the `_scalar_bar_actors` attribute of `BasePlotter` and are not linked to any particular mesh but to a data array name. This allows multiple meshes that are displaying the same data, e.g. "Density", to share one single scalar bar whose range spans the ranges across all meshes.

It may be worth looking into how we manage the `_scalar_bar_actors`, `_scalar_bar_mappers`, and `_scalar_bar_ranges`.

**Comment by math-artist** [comment]
Hey @banesullivan , thanks for your answer.

I don't know if this is a bug, but I can easily get a scalar bar displayed and at the same time have this during debug:
```
self.vtk_widget._scalar_bar_actors
Out[2]: {}
self.vtk_widget._scalar_bar_mappers
Out[3]: {}
self.vtk_widget._scalar_bar_ranges
Out[4]: {}
```

and this is a sample of what I use to add colored meshes or put them back to their initial display:
```
        s = something_in_3d

        if color_map is not None:
            scale = s.scales[color_map]
            if scalars is None:
                scalars = s.scalars
            actor = self.vtk_widget.add_mesh(s.mesh, None, scalars=scalars, clim=scale.clim, cmap=scale.cmap,
                                             label=None, reset_camera=reset_camera, scalar_bar_args=scale.settings,
                                             show_scalar_bar=True, name=s.name, annotations=scale.annotations,
                                             stitle="")
        else:
            actor = self.vtk_widget.add_mesh(s.mesh, s.color, label=None, reset_camera=reset_camera, show_scalar_bar=False,
                                             name=s.name, rgb=True)
            self.vtk_widget.remove_scalar_bar()
        if prop is not None:
            actor.SetProperty(prop)
        return actor
```
Even after a fresh add_mesh with a color scale, I get:

```
self.vtk_widget._scalar_bar_actors
Out[2]: {} 
```
right after the self.vtk_widget.add_mesh method. Could it be that my way of setting scalars directly doesn't track the scale bars?

**Comment by banesullivan** [comment]
Is that `if` statement getting executed? Because the `else` has `self.vtk_widget.remove_scalar_bar()` which would remove it?

**Comment by math-artist** [comment]
yes, I also tried it directly in debug without success. I have a toggle button that once clicked alternate between the if and else. The mesh is updated but not the scale.

Right now `self.vtk_widget.remove_scalar_bar()` only removes the scale bar after only importing one mesh. If I try to hide and restore the mesh, or if I add another mesh, that statement doesn't work.

I am currently working on putting my scalar arrays in the mesh instead of specifying directly. I have seen extra code in pyvista when the input scalars argument is a string.

**Comment by math-artist** [comment]
I found the culprit.
I use "" for titles because I don't want to display the name on screen.
The statement if title: is then false even if I have set it to "".

Would it be possible to differentiate the "visible title" and the name recorded for the scale_bar? For example, use the array name instead.

**Comment by math-artist** [comment]
Update: I tried to use " " instead since it is True, but I got another error while deleting the scale_bar:
```
    self.vtk_widget.remove_actor(name)
  File "C:\Users\Mak\Anaconda3\envs\fresh\lib\site-packages\pyvista\plotting\plotting.py", line 1594, in remove_actor
    renderer.remove_actor(actor, reset_camera)
  File "C:\Users\Mak\Anaconda3\envs\fresh\lib\site-packages\pyvista\plotting\renderer.py", line 628, in remove_actor
    _remove_mapper_from_plotter(self.parent, actor, False)
  File "C:\Users\Mak\Anaconda3\envs\fresh\lib\site-packages\pyvista\plotting\renderer.py", line 859, in _remove_mapper_from_plotter
    slot = plotter._scalar_bar_slot_lookup.pop(name)
KeyError: ' '
```
I have searched for why plotter._scalar_bar_slot_lookup is empty, then found only one place where it is set:

```
        # Automatically choose location if not specified
        if position_x is None or position_y is None:
            try:
                slot = min(self._scalar_bar_slots)
                self._scalar_bar_slots.remove(slot)
                self._scalar_bar_slot_lookup[title] = slot
```
Is it possible that deleting a scale bar works only when not specifying the location of the scale bar? I have temporarely removed my position_x and position_y settings to bypass the bug.
